### PR TITLE
Added Share Serialization and Created Share Struct

### PR DESCRIPTION
**Pull request by zer0x64**
The goal of this PR is to add serialization support to the library so that we can convert a share to and from a byte array easily. This way, shares can be stored, transmitted or even printed, depending on the use case.

To do this, I had to create a struct for a share, instead of the tuple that was used.

Let me know what you think about it and if you want me to change something!

### CODE REVIEW

**Comment by zer0x64** [comment]
On hold, something happened during commit and the code is broken, trying to fix it now.

**Comment by zer0x64** [comment]
Fixed, I was missing a semicolon somewhere.

**Comment by zer0x64** [comment]
One issue with this is that the threshold is not encoded with the share, as with "standard" implementations of Shamir, which means that we must know the threshold to recover the secret. I don't really see how to fix this with the current architecture. What do you think about it?

**Comment by c0dearm** [comment]
Wow! Thank you very much for your contribution! I am checking right now your code from my phone, which is of course not ideal, but I love what I am seeing so far :)

Tomorrow morning (it's 12 AM here now) I will pull your code and take a better look at it, I will let you know if there is something odd, but it looks promising!

Indeed, shares serialzation was something todo and also it makes the code more readable, I am astonished how fast you picked up with the current code.

Thank you very much!

**Comment by c0dearm** [comment]
Hi @zer0x64 ! I just pushed a commit with mostly a cleanup of the doc tests, I also included unit and benchmark tests for the new functionality. Please, let me know if there is something not clear :)

File: src/lib.rs
Comment (c0dearm) [inline src/lib.rs:85]: Here I used `HashMap` to enforce all shares to be different. If for example, the minimum threshold is 3, but we have two duplicated shares, the password is still irrecoverable, even if the length of the list was 3.

We should discuss if changing to a slice is beneficial, or we should go back to using the `HashMap`, because right now, the check `shares.len() < self.0` is somewhat useless.

**Review by c0dearm — CHANGES_REQUESTED** [review]
@zer0x64 please check my single comment, I want to discuss this before approving :smiley:

File: src/lib.rs
Comment (zer0x64) [reply src/lib.rs:85]: Oh yeah, I see it, that's something I didn't catch!

One naive way to do this would be to redo the length check manually(i could easily write a small algorithm that checks for X duplicates, it isn't really hard), but that might incur a small performance drop compared to the HashMap.

Another way would be to make a HashSet from the `x` part and check the length of that. I believe that would be simpler and more effective that even the HashMap.

One thing I don't really like about directly sending an HashMap to the public API is simply about the complexity for the user. It is a bit harder to use for the user than a simple slice, and also you might want tit to be easy to use even for people who don't understand the math behind it. If you accept a slice, it abstracts to the user of the library the coordinate aspect of the share.

In the end that's your call, but I do believe the best would be to replace the `shares.len() < self.0` by a similar check but on a HashSet. I can commit the change, it's not a problem!

Comment (c0dearm) [reply src/lib.rs:85]: Great! I agree with you, let's put the `x` values in a `HashSet` and compare its length 👍 

Please, also add a unit test to check it works. Many thanks!

Comment (zer0x64) [reply src/lib.rs:85]: What I'm suggesting is this(works as expected with my manual tests):
```rust
pub fn recover(&self, shares: &[Share]) -> Result<Vec<u8>, &str> {
    let shares_x: HashSet<u8> = HashSet::from(shares.iter().map(|s| s.x.0).collect());

    if shares_x.len() < self.0 as usize {
        Err("Not enough shares to recover original secret")
    } else {
        Ok(math::interpolate(shares))
    }
}
````

Comment (zer0x64) [reply src/lib.rs:85]: Done and committed!

Comment (c0dearm) [reply src/lib.rs:85]: Perfect! I see you didn't use `from` in the end, that would have added overhead :)

I am going to approve your PR :heart:

Comment (zer0x64) [reply src/lib.rs:85]: Clippy warned me this was useless! Didn't know I could simply collect into a HashSet!

**Review by c0dearm — APPROVED** [review]

**Comment by zer0x64** [comment]
Never really dealt with commit signing, do I have to do anything about it?

**Comment by c0dearm** [comment]
Not necessary, I disabled it. I had to do it once for the `helm/charts` project and it is a pain in the ass :see_no_evil:

**Pull request merged**
